ratis-server/src/main/java/org/apache/ratis/server/impl/RaftServerImpl.java

```diff
@@ -1087,7 +1090,8 @@ private CompletableFuture<RaftClientReply> readAsync(RaftClientRequest request)
 
       return replyFuture
           .thenCompose(readIndex -> getReadRequests().waitToAdvance(readIndex))
-          .thenCompose(readIndex -> queryStateMachine(request))
+          .thenComposeAsync(ignored -> stateMachine.query(request.getMessage()), readAsyncExecutor)
```

Comment:
@symious , Since it is calling stateMachine.query(..), the `StateMachine` implementation could choose to run it with its own executor.  If we use an executor here, then the stateMachine has no choice and is forced to run it with Ratis executor.  So I suggest not to make this change.